File: backend/src/main/java/com/plugga/backend/dao/DeckCardDAOImpl.java
package com.plugga.backend.dao;

import com.plugga.backend.entity.DeckCard;
import org.hibernate.Session;
import org.hibernate.query.Query;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import java.util.List;

@Repository
public class DeckCardDAOImpl implements DeckCardDAO {
    private EntityManager entityManager;

    @Autowired
    public DeckCardDAOImpl(final EntityManager entityManager) {
        this.entityManager = entityManager;
    }

    @Override
    public List<DeckCard> findAll() {
        Session currentSession = entityManager.unwrap(Session.class);
        Query<DeckCard> query = currentSession.createQuery("from DeckCard", DeckCard.class);
        return query.getResultList();
    }

    @Override
    public DeckCard findById(final int id) {
        Session currentSession = entityManager.unwrap(Session.class);
        return currentSession.get(DeckCard.class, id);
    }

    @Override
    public List<DeckCard> findByDeckId(final int deckId) {
        Session currentSession = entityManager.unwrap(Session.class);
        Query<DeckCard> query = currentSession.createQuery("from DeckCard where deck=:deckId", DeckCard.class);
        query.setParameter("deckId", deckId);
        return query.getResultList();
    }

    @Override
    public List<DeckCard> findByCardId(final int cardId) {
        Session currentSession = entityManager.unwrap(Session.class);
        Query<DeckCard> query = currentSession.createQuery("from DeckCard where card=:cardId", DeckCard.class);
        query.setParameter("cardId", cardId);
        return query.getResultList();
    }

    @Override
    public void save(final DeckCard deckCard) {
        Session currentSession = entityManager.unwrap(Session.class);
        currentSession.saveOrUpdate(deckCard);
    }

    @Override
    public void deleteById(final int id) {
        Session currentSession = entityManager.unwrap(Session.class);
        Query query = currentSession.createQuery("delete from DeckCard where id=:deckCardId");
        query.setParameter("deckCardId", id);
        query.executeUpdate();
    }
}

File: package/src/main/kotlin/com/plugga/backend/dao/DeckCardDAOImpl.kt
package com.plugga.backend.dao

import com.plugga.backend.entity.DeckCard
import org.hibernate.Session
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Repository

import javax.persistence.EntityManager

@Repository
class DeckCardDAOImpl @Autowired
constructor(private val entityManager: EntityManager) : DeckCardDAO {

    override fun findAll(): MutableList<DeckCard> {
        val currentSession = entityManager.unwrap(Session::class.java)
        val query = currentSession.createQuery("from DeckCard", DeckCard::class.java)
        return query.resultList
    }

    override fun findById(id: Int): DeckCard? {
        val currentSession = entityManager.unwrap(Session::class.java)
        return currentSession.get(DeckCard::class.java, id)
    }

    override fun findByDeckId(deckId: Int): MutableList<DeckCard> {
        val currentSession = entityManager.unwrap(Session::class.java)
        val query = currentSession.createQuery("from DeckCard where deck_id=:deckId", DeckCard::class.java)
        query.setParameter("deckId", deckId)
        return query.resultList
    }

    override fun findByCardId(cardId: Int): MutableList<DeckCard> {
        val currentSession = entityManager.unwrap(Session::class.java)
        val query = currentSession.createQuery("from DeckCard where card_id=:cardId", DeckCard::class.java)
        query.setParameter("cardId", cardId)
        return query.resultList
    }

    override fun save(deckCard: DeckCard) {
        val currentSession = entityManager.unwrap(Session::class.java)
        currentSession.saveOrUpdate(deckCard)
    }

    override fun deleteById(id: Int) {
        val currentSession = entityManager.unwrap(Session::class.java)
        val query = currentSession.createQuery("delete from DeckCard where id=:deckCardId")
        query.setParameter("deckCardId", id)
        query.executeUpdate()
    }
}
